### PR TITLE
uttmm - 17 Edit Project, uttmm - 25 List of all projects, uttm - 37 Pagination Projects, uttmm - 45 - Get all api for project types

### PR DESCRIPTION
**Pull request by stetskivostap**

### CODE REVIEW

File: data/Piranha.Data.EF/Db.cs [hunk 1]
```diff
@@ -155,7 +155,7 @@ public Db(DbContextOptions<Db> options) : base(options)
                     if (!IsInitialized)
                     {
                         // Migrate database
-                        Database.Migrate();
+                        //Database.Migrate();
```
Comment (dgzargo) [inline data/Piranha.Data.EF/Db.cs]: why did you do this?

File: examples/MvcWeb/TheraLang/Controllers/PaginationController.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    {
+        IttmmDbContext context;
+        public PaginationController(IttmmDbContext Context)
+        {
+            this.context = Context;
+        }
+        [HttpGet]
+        public async Task<IActionResult> Index(int page = 1)
+        {
+            int pageSize = 5;
```
Comment (dgzargo) [inline examples/MvcWeb/TheraLang/Controllers/PaginationController.cs]: hardcoded page size

File: examples/MvcWeb/TheraLang/Controllers/PaginationController.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+using MvcWeb.TheraLang.Entities;
+using MvcWeb.TheraLang.Pagination;
+namespace MvcWeb.TheraLang.Controllers
+{
+    [Route("api/pag/projects")]
+    [ApiController]
+    public class PaginationController : Controller
+    {
+        IttmmDbContext context;
+        public PaginationController(IttmmDbContext Context)
```
Comment (dgzargo) [inline examples/MvcWeb/TheraLang/Controllers/PaginationController.cs]: don't use DbContext in controller, please!

Comment (stetskivostap) [reply examples/MvcWeb/TheraLang/Controllers/PaginationController.cs]: ohh sry,  dsdn`t notice

File: examples/MvcWeb/TheraLang/Controllers/PaginationController.cs (hunk 2 above)
Comment (stetskivostap) [reply examples/MvcWeb/TheraLang/Controllers/PaginationController.cs]: what exactly are u talking about? If we didn`t send number of page explicitly, it goes on page 1,
about page size? - i can change it

Comment (stetskivostap) [reply examples/MvcWeb/TheraLang/Controllers/PaginationController.cs]: > hardcoded page size

what exactly are u talking about? If we didn`t send number of page explicitly, it goes on page 1,
about page size? - i can change it

File: examples/MvcWeb/TheraLang/Controllers/PaginationController.cs (hunk 3 above)
Comment (stetskivostap) [reply examples/MvcWeb/TheraLang/Controllers/PaginationController.cs]: > don't use DbContext in controller, please!

ohh sry, dsdn`t notice

File: data/Piranha.Data.EF/Db.cs (hunk 1 above)
Comment (stetskivostap) [reply data/Piranha.Data.EF/Db.cs]: > why did you do this?

I showed  you that bug, so decided to comment

Comment (yrikturash) [reply data/Piranha.Data.EF/Db.cs]: you mean not working, okay, we need to solve it - will add to todo list

Comment (yrikturash) [reply data/Piranha.Data.EF/Db.cs]: please add todo comment there, like
//TODO: fix migrations

File: examples/MvcWeb/TheraLang/Controllers/PaginationController.cs (hunk 2 above)
Comment (yrikturash) [reply examples/MvcWeb/TheraLang/Controllers/PaginationController.cs]: he means that you should move it to Constans file, please add file Constants.cs into MVCWeb -> TheraLang folder and code inside
`public static class PagingConstants
    {
        public const int StartPage= 1;
        public const int ItemsPerPage = 10;
    }`

File: examples/MvcWeb/TheraLang/Controllers/ProjectTypeController.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+            return service.GetAllProjectsType();
+        }
+        [HttpGet("{id}")]
+        public IActionResult GetType(int id)
+        {
+            var type = GetAllTypes().FirstOrDefault(p => p.Id == id);
+            if (type == null)
+            {
+                throw new NullReferenceException($"type with {id} not found");
+            }
```
Comment (yrikturash) [inline examples/MvcWeb/TheraLang/Controllers/ProjectTypeController.cs]: you should create method for this in the ProjectTypeService

File: examples/MvcWeb/TheraLang/Services/ProjectTypeService.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    public class ProjectTypeService : IProjectTypeService
+    {
+        private IUnitOfWork uow { get; }
+        public ProjectTypeService(IUnitOfWork unitOfWork)
+        {
+            uow = unitOfWork;
+        }
+        public IEnumerable<ProjectType> GetAllProjectsType()
+        {
+            return uow.Repository<ProjectType>().Get().ToList();
```
Comment (yrikturash) [inline examples/MvcWeb/TheraLang/Services/ProjectTypeService.cs]:
```suggestion
            return uow.Repository<ProjectType>().Get().AsNoTracking().ToList();
```

File: examples/MvcWeb/TheraLang/Controllers/ProjectTypeController.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        {
+            service = Service;
+        }
+        [HttpGet]
+        public IEnumerable<ProjectType> GetAllTypes()
+        {
+            return service.GetAllProjectsType();
+        }
+        [HttpGet("{id}")]
+        public IActionResult GetType(int id)
```
Comment (yrikturash) [inline examples/MvcWeb/TheraLang/Controllers/ProjectTypeController.cs]:
```suggestion
        public IActionResult GetType(int id){
        if(id == default(int)){
            new ArgumentException($"{nameof(id)} cannot be 0");
            }
        }
```

File: examples/MvcWeb/TheraLang/Controllers/ProjectController.cs [hunk 7]
```diff
@@ -1,35 +1,58 @@
...
         public IActionResult GetProject(int id)
         {
-            var project = GetAllProjects().FirstOrDefault((p) => p.Id == id);
-            if (project == null)
+           var project = ProjectService.GetById(id);
+            if(project == null)
             {
-                return NotFound();
+                throw new System.NullReferenceException($"project with id {id} not found");
             }
```
Comment (yrikturash) [inline examples/MvcWeb/TheraLang/Controllers/ProjectController.cs]: remove - exists on service

File: examples/MvcWeb/TheraLang/Controllers/ProjectController.cs [hunk 8]
```diff
@@ -1,35 +1,58 @@
...
+            return Ok(project);            
+        }
+        [HttpPut("update{id}")]
+        public IActionResult EditProject(int id,Project project)
+        {
+            ProjectService.UpdateAsync(id,project);
+            if (project == null)
+            {
+                throw new System.NullReferenceException($"project with id {id} not found");
+            }
```
Comment (yrikturash) [inline examples/MvcWeb/TheraLang/Controllers/ProjectController.cs]: remove - exists on service

File: examples/MvcWeb/TheraLang/Controllers/ProjectController.cs [hunk 9]
```diff
@@ -1,35 +1,58 @@
...
 namespace MvcWeb.TheraLang.Controllers
 {
     [Route("api/projects")]
     [ApiController]
     public class ProjectController : ControllerBase
     {
-
-        public IUnitOfWork uow;
-        public ProjectController(IUnitOfWork unitOfWork)
+        public IProjectService ProjectService;
```
Comment (yrikturash) [inline examples/MvcWeb/TheraLang/Controllers/ProjectController.cs]:
```suggestion
        private readonly IProjectService _projectService; 
```

File: examples/MvcWeb/TheraLang/Controllers/ProjectController.cs [hunk 10]
```diff
@@ -1,35 +1,58 @@
-﻿using System.Collections.Generic;
+
```
Comment (yrikturash) [inline examples/MvcWeb/TheraLang/Controllers/ProjectController.cs]: remove

**Comment by yrikturash** [comment]
Add uploads folder with images to ignore git file

**Pull request closed without merging**